enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,737 @@
+
+
+### Goals
+
+Provide guidance and example artifacts for configuring the system for low latency workload running on MicroShift:
+- Prepare low latency TuneD profile for MicroShift
+- Prepare necessary CRI-O configurations
+- Allow configuration of Kubelet via MicroShift config
+- Introduce a mechanism to automatically apply a tuned profile upon boot.
+- Document how to create a new tuned profile for users wanting more control.
+
+
+### Non-Goals
+
+- Workload partitioning (i.e. pinning MicroShift control plane components) (see [OCPSTRAT-1068](https://issues.redhat.com/browse/OCPSTRAT-1068))
+- Duplicate all capabilities of Node Tuning Operator
+
+
+## Proposal
+
+To ease configuration of the system for running low latency workloads on MicroShift following
+parts need to be put in place:
+- `microshift-baseline` TuneD profile
+- CRI-O configuration + Kubernetes' RuntimeClass
+- Kubelet configuration (CPU, Memory, and Topology Managers and other)
+- `microshift-tuned.service` to activate user selected TuneD profile on boot and reboot the host
+  if the kernel args are changed.
+
+New `microshift-low-latency` RPM will be created that will contain tuned profile, CRI-O configs, and mentioned systemd daemon.
+We'll leverage existing know how of Performance and Scalability team expertise and look at
+Node Tuning Operator capabilities.
+
+To allow customization of supplied TuneD profile for specific system, this new profile will
+include instruction to include file with variables, which can be overridden by the user.
+
+All of this will be accompanied by step by step documentation on how to use this feature,
+tweak values for specific system, and what are the possibilities and limitations.
+
+Optionally, a new subcommand `microshift doctor low-latency` might be added to main
+MicroShift binary to provide some verification checks if system configuration is matching
+expectations according to our knowledge. It shall not configure system - only report potential problems.
+
+
+### Workflow Description
+
+Workflow consists of two parts:
+1. system and MicroShift configuration
+1. preparing Kubernetes manifests for low latency
+
+#### System and MicroShift configuration
+
+##### OSTree
+
+1. User creates an osbuild blueprint:
+   - (optional) User configures `[customizations.kernel]` in the blueprint if the values are known
+     beforehand. This could prevent from necessary reboot after applying tuned profile.
+   - (optional) User adds `kernel-rt` package to the blueprint
+   - User adds `microshift-low-latency.rpm` to the blueprint
+   - User enables `microshift-tuned.service`
+   - User supplies additional configs using blueprint:
+     - `/etc/tuned/microshift-baseline-variables.conf` to configure tuned profile
+     - `/etc/microshift/config.yaml` to configure Kubelet
+     - `/etc/microshift/microshift-tuned.yaml` to configure `microshift-tuned.service`
+1. User builds the blueprint
+1. User deploys the commit / installs the system.
```

Comment:
Is there any way to verify we did what we think we did before restarting?

Reply:
Is this question for right line?
In general there is `tuned-adm verify` command which compares system's desired state with current state but it works weirdly on ostree systems (complains about some things) whereas on RPM systems the verification is passing.
So in this case, successful execution of `tuned-adm profile PROFILE-NAME` should be sufficient.